File: examples/receipt_line.py
"""Append receipt lines for Agent Receipt. Copy this file into your agent.

    from receipt_line import record
    record("send_email", target="a@b.c", agent="my-bot", id="msg_1", reversible=False)

Writes one JSON line per call to ~/.agent-receipt/inbox/<agent>.jsonl.
Format: docs/RECEIPT_LINE.md. Stdlib only.
"""

import json
import os
import re
import time
from datetime import datetime, timezone

INBOX = os.environ.get("AGENT_RECEIPT_INBOX") or os.path.expanduser("~/.agent-receipt/inbox")
ACTIONS = {"send_email", "create_event", "purchase", "file_write", "post", "execute", "other"}


def record(action, target=None, agent="agent", id=None, amount=None, currency=None,
           link=None, reversible=None, reason=None, session=None, cwd=None, project=None,
           detail=None, ts=None):
    if action not in ACTIONS:
        raise ValueError(f"action must be one of {sorted(ACTIONS)}")
    line = {
        "ts": ts or datetime.now(timezone.utc).isoformat(timespec="milliseconds"),
        "agent": agent,
        "action": action,
    }
    for key, value in (("target", target), ("id", id), ("amount", amount), ("currency", currency),
                       ("link", link), ("reversible", reversible), ("reason", reason),
                       ("session", session), ("cwd", cwd), ("project", project), ("detail", detail)):
        if value is not None:
            line[key] = value
    os.makedirs(INBOX, exist_ok=True)
    name = re.sub(r"[^A-Za-z0-9_.-]+", "_", agent) or "agent"
    with open(os.path.join(INBOX, f"{name}.jsonl"), "a", encoding="utf-8") as f:
        f.write(json.dumps(line, ensure_ascii=False) + "\n")
    return line


if __name__ == "__main__":
    print(record("other", target="self-test", agent="receipt-line-example", id=f"selftest-{int(time.time())}"))

File: examples/receipt_line.js
// Append receipt lines for Agent Receipt. Copy this file into your agent.
//
//   const { record } = require("./receipt_line");
//   record("purchase", { target: "AWS", amount: 42.1, currency: "USD", agent: "ops-bot", id: invoiceId });
//
// Writes one JSON line per call to ~/.agent-receipt/inbox/<agent>.jsonl.
// Format: docs/RECEIPT_LINE.md. No dependencies.

const fs = require("fs");
const os = require("os");
const path = require("path");

const INBOX = process.env.AGENT_RECEIPT_INBOX || path.join(os.homedir(), ".agent-receipt", "inbox");
const ACTIONS = new Set(["send_email", "create_event", "purchase", "file_write", "post", "execute", "other"]);

function record(action, fields = {}) {
  if (!ACTIONS.has(action)) throw new Error(`action must be one of ${[...ACTIONS].join(", ")}`);
  const agent = fields.agent || "agent";
  const line = { ts: fields.ts || new Date().toISOString(), agent, action };
  for (const key of ["target", "id", "amount", "currency", "link", "reversible", "reason", "session", "cwd", "project", "detail"]) {
    if (fields[key] !== undefined && fields[key] !== null) line[key] = fields[key];
  }
  fs.mkdirSync(INBOX, { recursive: true });
  const name = agent.replace(/[^A-Za-z0-9_.-]+/g, "_") || "agent";
  fs.appendFileSync(path.join(INBOX, `${name}.jsonl`), JSON.stringify(line) + "\n");
  return line;
}

module.exports = { record };

if (require.main === module) {
  console.log(record("other", { target: "self-test", agent: "receipt-line-example", id: `selftest-${Date.now()}` }));
}
